### PR TITLE
Use the API to install repositories instead of loading the toolshed in an iframe.

### PR DESCRIPTION
**Pull request by davebx**
There remains work to be done on this, I'm sure the javascript in preview_repository.mako could be dramatically improved, but the current code should be functional.

### CODE REVIEW

**Comment by davebx** [comment]
Pings to @erasche, @jmchilton, @blankenberg, @bgruening, and @martenson. This might be relevant to your interests.

File: lib/galaxy/webapps/tool_shed/api/categories.py [hunk 1]
```diff
@@ -80,6 +83,7 @@ def index( self, trans, deleted=False, **kwd ):
             category_dict[ 'url' ] = web.url_for( controller='categories',
                                                   action='show',
                                                   id=trans.security.encode_id( category.id ) )
+            category_dict[ 'repositories' ] = trans.app.repository_registry.viewable_repositories_and_suites_by_category.get( category.name, 0 )
```
Comment (martenson) [inline lib/galaxy/webapps/tool_shed/api/categories.py]: I think this might want to have its own endpoint at e.g. `/api/categories/<id>/repositories` to keep requests atomic.

File: lib/galaxy/webapps/tool_shed/api/categories.py [hunk 2]
```diff
@@ -103,4 +107,6 @@ def show( self, trans, id, **kwd ):
         category_dict[ 'url' ] = web.url_for( controller='categories',
                                               action='show',
                                               id=trans.security.encode_id( category.id ) )
+        if util.asbool( kwd.get( 'show_repositories', False ) ):
+            category_dict[ 'repositories' ] = suc.get_repositories_by_category( trans.app, category.id )
```
Comment (martenson) [inline lib/galaxy/webapps/tool_shed/api/categories.py]: similar to L86 above - this is better but separate endpoint would be the best I believe

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py [hunk 3]
```diff
@@ -202,6 +231,47 @@ def index( self, trans, **kwd ):
...
+        changeset = kwd.get( 'changeset', None )
+        tool_shed_url = kwd.get( 'tool_shed_url', None )
+        params[ 'tool_panel_section_id' ] = kwd.get( 'tool_panel_section_id', None )
+        params[ 'install_tool_dependencies' ] = util.asbool( kwd.get( 'install_tool_dependencies', False ) )
+        params[ 'install_repository_dependencies' ] = util.asbool( kwd.get( 'install_repository_dependencies', False ) )
+        params[ 'shed_tool_conf' ] = kwd.get( 'shed_tool_conf', None )
+        params[ 'tool_path' ] = suc.get_tool_path_by_shed_tool_conf_filename( trans.app, params[ 'shed_tool_conf' ] )
+        async = util.string_as_bool( kwd.get( 'async', True ) )
+        # def install( self, tool_shed_url, name, owner, changeset_revision, install_options ):
+        if async:
```
Comment (hexylena) [inline lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py]: async seems to be a hard required parameter. Document it?

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py [hunk 4]
```diff
@@ -202,6 +231,47 @@ def index( self, trans, **kwd ):
         return tool_shed_repository_dicts
 
     @expose_api
+    def install( self, trans, payload, **kwd ):
+        """
+        POST /api/tool_shed_repositories/install
+        Initiate the installation of one or more repositories.
+
+        :param encoded_kwd (required): The encoded dict describing the repositories to be installed.
```
Comment (hexylena) [inline lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py]: encoded_kwd isn't used in this function ... the docs don't really line up with the code below.

File: lib/galaxy/webapps/tool_shed/api/repositories.py [hunk 5]
```diff
@@ -86,8 +86,36 @@ def add_repository_registry_entry( self, trans, payload, **kwd ):
             % ( name, owner )
         return response_dict
 
+    @expose_api_anonymous_and_sessionless
+    def get_latest_metadata( self, trans, id, **kwd ):
```
Comment (jmchilton) [inline lib/galaxy/webapps/tool_shed/api/repositories.py]: Is this related to the PR, how is it consumed if it is?

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py [hunk 6]
```diff
@@ -378,20 +448,20 @@ def repair_repository_revision( self, trans, payload, **kwd ):
...
+        owner = payload.get( 'owner', None )
+        if tool_shed_url is None:
             raise exceptions.RequestParameterMissingException( "Missing required parameter 'tool_shed_url'." )
-        name = payload.get( 'name', '' )
-        if not name:
+        if name is None:
             raise exceptions.RequestParameterMissingException( "Missing required parameter 'name'." )
-        owner = payload.get( 'owner', '' )
-        if not owner:
+        if owner is None:
```
Comment (hexylena) [inline lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py]: are empty names permitted? If not, then we should probably retain the original code here or throw a different exception

**Comment by hexylena** [comment]
Really exciting work @davebx :D

File: lib/galaxy/webapps/tool_shed/api/categories.py (hunk 1 above)
Comment (davebx) [inline lib/galaxy/webapps/tool_shed/api/categories.py]: As discussed offline, this one only shows the count of repositories in the relevant category.

File: lib/galaxy/webapps/tool_shed/api/repositories.py (hunk 5 above)
Comment (davebx) [inline lib/galaxy/webapps/tool_shed/api/repositories.py]: Reposting explanation I provided offline:

This endpoint is used to populate the "tools or package verified" column in the browse category page.

**Comment by bgruening** [comment]
Great work Dave!
@galaxybot test this

**Comment by davebx** [comment]
After the changes in 8d2e173, the interface looks like:

Screenshot as of 8d2e173

**Comment by bgruening** [comment]
Like it!

**Comment by davebx** [comment]
@bgruening Naturally I chose the most complicated dependency structure I could think of for the almost-ready-to-remove-the-wip-tag example.

**Comment by bgruening** [comment]
@davebx what is needed to get this out of WIP?
Tests are passing I think it would be good to get this into dev as soon as possible.

If you like I can create a Docker container of this branch to make testing easier.

ping @gregvonkuster you might be interested in reviewing.

**Comment by davebx** [comment]
@bgruening The search field needs to become more functional. I asked @martenson to help out with that, but his plate is full right now. And naturally @gregvonkuster's input would be highly useful.

File: lib/galaxy/web/form_builder.py [hunk 7]
```diff
@@ -312,7 +313,11 @@ def __init__( self, name, multiple=None, display=None, refresh_on_change=False,
     def add_option( self, text, value, selected=False ):
         self.options.append( ( text, value, selected ) )
 
-    def get_html( self, prefix="", disabled=False ):
+    def get_html( self, prefix="", disabled=False, extra_attr=None ):
+        if extra_attr is not None:
+            self.extra_attributes = ' %s' % ' '.join( [ '%s="%s"' % ( k, v ) for k, v in extra_attr.items() ] )
```
Comment (jmchilton) [inline lib/galaxy/web/form_builder.py]: Can you attribute escape this (https://docs.python.org/2/library/xml.sax.utils.html#xml.sax.saxutils.quoteattr) - I guess it is only used in one place so it isn't currently exposing a vulnerability but I'd like other potential consumers of this to have their values automatically escaped - at least by default.

File: lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py [hunk 8]
```diff
@@ -41,7 +42,7 @@ class ToolShedRepositoriesController( BaseAPIController ):
 
     def __ensure_can_install_repos( self, trans ):
         # Make sure this Galaxy instance is configured with a shed-related tool panel configuration file.
-        if not suc.have_shed_tool_conf_for_install( trans.app ):
+        if not suc.have_shed_tool_conf_for_install( self.app ):
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py:42]: Awesome, thanks!

**Comment by jmchilton** [comment]
The code and idea look great to me - so :+1: - but I haven't tested it. Ideally a tool shed fan and user would do that - @martenson, @bgruening?

**Comment by bgruening** [comment]
To make testing easier:

```
docker run -p 8080:80 quay.io/bgruening/galaxy:toolshed_decouple
```

Update: I see this does not work as expected as API changes and functionality is in the same PR. 
@davebx so the only way to test this is with setting up an own ToolShed + complex repositories like the CTB?

If so can we split API changes and the TS changes into separate PR and get the test toolshed updated ;)
Otherwise lets merge it, we are in the beginning of this release schedule and this will not break the old TS install process, it will add a beta version.

I'm +1 for this one.

**Comment by davebx** [comment]
@bgruening I think it would be fairly easy to split the toolshed and galaxy parts into separate pull requests.

**Comment by bgruening** [comment]
Depends on you, I'm actually fine with merging and trusting you to fix problems as they appear until the next release.

It's just hard to bootstrap a TS with a complex repo dependency and the TTS would make it much easier!

ping @martenson, @jmchilton ?

**Comment by martenson** [comment]
I am testing this now, some findings:
(not all of these are  necessary for merge, some features will come on next phase)
- [x] When opening a simple repo I will get the following JS error:
  `Uncaught TypeError: Cannot read property 'has_repository_dependencies' of undefined`
- [x] JS logging at /browse_toolshed is ON
- [x] /preview repository will yield JS error `Uncaught TypeError: Cannot read property 'length' of undefined`
- [x] panel sections to choose from is empty for me for tool without dependencies, possibly due to the JS errors above
- [x] revisions select is empty for tool without dependencies, possibly same error as above
- [x] show the label of item when searching, not the value (repo id)
- [ ] search needs to accept IDs
- [x] offer search on category browse page
- [x] if a repository does not have tools, the valid tools section is empty (but it should be hidden completely)
- [ ] the search field displays some revisions (91c89b6f01b9f2ae) during search, overall this feel a little bit clumpy

**Comment by davebx** [comment]
Also, @bgruening, you're welcome to configure my work desktop as a toolshed for your galaxy, it's kept current with davebx/galaxy/toolshed_decouple or decouple_api.

**Comment by bgruening** [comment]
I will add my findings to the checkbox from Martin.

**Comment by bgruening** [comment]
Actually all tools beside the meta-repository `chemicaltoolbox` does not have a revision and a valid tools. This can be due to the test tool shed. Bus this is strange. @davebx this is with your TS and the Docker container.

**Comment by davebx** [comment]
There were some issues importing the CTB capsule, that may be related.

**Comment by carlfeberhard** [comment]
Heya, @davebx like we talked about I'll comment on the JS stuff since you thought it would help. The actual functionality here I'll leave to the experts on tool installation (including you).

On the JS, overall it's great. Here are some changes I'd strongly encourage:
- don't override Array's prototype - it's not _always_ bad, but it can lead to problems for other scripts on the page or when it's incorporated with other scripts. I'd suggest a function with the array as an arg or an underscore function if you find one that will will work
- use `===` instead of `==` and `!==` instead of `!=`: the two character versions can do some casting between operands that might be unexpected
- use mako to add scripts to the head if possible (inline note on `preview_repository.mako`)

Here's a good guide that I like that might help: https://github.com/bevacqua/js
There are other good guides out there too.

Nitpicking (up to you):
- I prefer tall dictionary/Object/map definitions instead of wide. In other words, having a single key/value on one line each if you've got more than three k/v pairs
- you can move the `_.template` calls out from the functions that render the templates. Currently, what you have will build the template each time the rendering function is called. If you move the calls out, they will be built only once when the script is evaluated. Good to see template being used, though

Damn impressive since you just started learning javascript.

**Comment by martenson** [comment]
@carlfeberhard awesome guide, thanks for sharing!

**Comment by davebx** [comment]
The input is appreciated, @carlfeberhard. Watch this space for commits changing the JS.

File: templates/admin/tool_shed_repository/preview_repository.mako [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+
+<%def name="stylesheets()">
+    ${parent.stylesheets()}
+    ${h.css( "dynatree_skin/ui.dynatree" )}
+</%def>
+
+<%def name="javascripts()">
+    ${parent.javascripts()}
+</%def>
+${javascripts()}
```
Comment (carlfeberhard) [inline templates/admin/tool_shed_repository/preview_repository.mako]: (Sorry, @davebx I forgot to add this part)
In order to have the script you're adding from line 19 to 339 appear in the head of the page, you can move the `</%def>` past 339. In any event, you don't need to call like 14 (it's called in base.mako IIRC) and don't need lines 11 to 13 at all if you're not adding anything to the parent's `javascripts()` call.

You can similarly move the styles defined at 339 into the stylesheets mako function on 6.

**Comment by davebx** [comment]
@martenson I believe I have corrected many of the points on your checklist.

**Comment by davebx** [comment]
The failing toolshed test is a result of the main toolshed being unavailable while @natefoo performs some much-needed maintenance.

**Comment by bgruening** [comment]
@galaxybot test this

**Comment by mvdbeek** [comment]
@davebx this is amazing, I'm testing it right now, and the UI is very nice!
One thing that I would do is to throw a nicer error message if targeting an older toolshed!

If more people want to test this, feel free point your galaxy instances at https://lbcd41.snv.jussieu.fr/toolshed ... though there are just a few repos to test. But I guess you can always import capsules?!

**Comment by mvdbeek** [comment]
Okay, this is seriously awesome. The bedtools suite installed fine, and in three different toolpanel (one default, and 2 in 2 seperate) sections. But it does log errors when there are actually no errors:

```
....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,130 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_bed12tobed6/2.24.0, version: 2.24.0 into tool panel....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,130 Error while loading tool lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/0cc0bc30f78c/bedtools/genomeCoverageBed.xml : 
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,167 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_genomecoveragebed/2.24.0, version: 2.24.0 into tool panel....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,167 Error while loading tool lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/0cc0bc30f78c/bedtools/spacingBed.xml : 
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,197 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_spacingbed/2.24.0, version: 2.24.0 into tool panel....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,197 Error while loading tool lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/0cc0bc30f78c/bedtools/expandBed.xml : 
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,229 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_expandbed/2.24.0, version: 2.24.0 into tool panel....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,229 Error while loading tool lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/0cc0bc30f78c/bedtools/bamToFastq.xml : 
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,259 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_bamtofastq/2.24.0, version: 2.24.0 into tool panel....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,260 Error while loading tool lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/0cc0bc30f78c/bedtools/annotateBed.xml : 
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,292 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_annotatebed/2.24.1, version: 2.24.1 into tool panel....
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,293 Error while loading tool lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/0cc0bc30f78c/bedtools/intersectBed.xml : 
galaxy.tools.toolbox.base DEBUG 2016-03-11 14:57:51,338 Loaded tool id: lbcd41.snv.jussieu.fr/toolshed/repos/iuc/bedtools/bedtools_intersectbed/2.24.0, version: 2.24.0 into tool panel....
```

**Comment by mvdbeek** [comment]
Okay, I see the error has nothing to do with your PR, but us running the [toolshed in a subdirectory](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/toolbox/base.py#L562).

I can't judge the quality of the javascript / mako part, but it does seem to work well. :+1:

**Comment by bgruening** [comment]
@davebx, @martenson can we merge this before feature-freeze. It's working for me and @mvdbeek and it does not break the old form so I consider this a save move.

**Comment by davebx** [comment]
@bgruening I'll update it to fix the merge conflict, in case your motion carries.

**Comment by martenson** [comment]
I am in favor of merging as long as it is clearly labeled (Beta) which it is and does not interfere with the current approach. 
:+1:

**Comment by bgruening** [comment]
@davebx are you working on the two missing bullet-points above? Should we migrate them to an issue and track the progress there?

**Comment by davebx** [comment]
@bgruening Yes, an issue for cleaning up the search feature would be a good idea.

**Comment by martenson** [comment]
@davebx @bgruening I will implement the two search points as a part of porting toolshed.tools search  here.

**Comment by mvdbeek** [comment]
Nice work @davebx !

**Comment by bgruening** [comment]
@martenson thanks!
Great work @davebx!
